reactor3/src/test/java/com/linecorp/armeria/common/reactor3/RequestContextPropagationMonoTest.java

```diff
@@ -0,0 +1,393 @@
+                            .verifyComplete();
+            }
+        } else {
+            StepVerifier.create(mono)
+                        .expectNextMatches(t -> "foo".equals(t.getT1()) && "bar".equals(t.getT2()))
+                        .verifyComplete();
+        }
+        assertThat(ctxExists(ctx)).isFalse();
+    }
+
+    @Test
+    void subscriberContextIsNotMissing() {
+        final ClientRequestContext ctx = newContext();
+        final Mono<String> mono;
+        mono = Mono.deferContextual(Mono::just).handle((reactorCtx, sink) -> {
+            assertThat((String) reactorCtx.get("foo")).isEqualTo("bar");
+            sink.next("baz");
+        });
+
+        final Mono<String> mono1 = mono.contextWrite(reactorCtx -> reactorCtx.put("foo", "bar"));
+        StepVerifier.create(mono1)
+                    .expectNextMatches("baz"::equals)
+                    .verifyComplete();
+        assertThat(ctxExists(ctx)).isFalse();
+    }
+
+    @Test
+    void ctxShouldBeCleanUpEvenIfErrorOccursDuringReactorOperationOnSchedulerThread()
+            throws InterruptedException {
+        // Given
+        final ClientRequestContext ctx = newContext();
+        final Mono<String> mono;
+        final Scheduler single = Schedulers.single();
+
+        // When
+        mono = Mono.just("Hello")
+                   .subscribeOn(single)
+                   .delayElement(Duration.ofMillis(1000))
+                   .map(s -> {
+                       if ("Hello".equals(s)) {
+                           throw new RuntimeException();
+                       }
+                       return s;
+                   })
+                   .contextWrite(Context.of(RequestContext.class, ctx));
+
+        // Then
+        StepVerifier.create(mono)
+                    .expectError(RuntimeException.class)
+                    .verify();
+
+        final CountDownLatch latch = new CountDownLatch(1);
+        single.schedule(() -> {
+            assertThat(ctxExists(ctx)).isFalse();
+            latch.countDown();
+        });
+        latch.await();
+
+        assertThat(ctxExists(ctx)).isFalse();
+    }
+
+    static Subscription noopSubscription() {
+        return new Subscription() {
+            @Override
+            public void request(long n) {}
+
+            @Override
+            public void cancel() {}
+        };
+    }
+
+    static boolean ctxExists(ClientRequestContext ctx) {
+        return RequestContext.currentOrNull() == ctx;
+    }
+
+    static ClientRequestContext newContext() {
+        return ClientRequestContext.builder(HttpRequest.of(HttpMethod.GET, "/"))
+                                   .build();
+    }
+
+    private static <T> Mono<T> addCallbacks(Mono<T> mono0, ClientRequestContext ctx,
+                                            boolean useContextCapture) {
+        // doOnCancel and doFinally do not have context because we cannot add a hook to the cancel.
```

Comment:
nit; I believe this isn't true anymore

```suggestion
```

We can also add the following callbacks for validation
```java
.doOnCancel(() -> assertThat(ctxExists(ctx)).isTrue())
.doFinally(t -> assertThat(ctxExists(ctx)).isTrue())
```